python/cudf/cudf/core/frame.py

```diff
@@ -42,129 +43,183 @@ def _concat(cls, objs, axis=0, ignore_index=False):
-                            )
-                            .codes
-                        )
-                    df[name] = df[name].astype(dtype)
-
-        # Arrange the columns from each DF in the same order
-        objs = [df[list(unique_columns_ordered_ls)] for df in objs]
+        cast_cols_to_common_dtypes(indices, columns, dtypes, categories)
+
+        # Construct input tables with the index and data columns in the same
+        # order. This strips the given index/column names and replaces the
+        # names with their integer positions in the `cols` list
+        tables = []
+        for i, cols in enumerate(columns):
+            table_cols = cols[first_data_column_position:]
+            table_names = indices[first_data_column_position:]
+            table = cls(data=dict(zip(table_names, table_cols)))
+            if 1 == first_data_column_position:
+                table._index = as_index(cols[0])
+            elif first_data_column_position > 1:
+                index_cols = cols[:first_data_column_position]
+                index_names = indices[:first_data_column_position]
+                table._index = cls(data=dict(zip(index_names, index_cols)))
+            tables.append(table)
 
         # Concatenate the Tables
-        out = cls(libcudfxx.concat.concat_tables(objs)._data, index)
+        out = cls._from_table(libcudfxx.concat.concat_tables(
+            tables, ignore_index=ignore_index
+        ))
+
+        # Reassign the categories for any categorical table cols
+        reassign_categories(
+            categories, out._data,
+            indices[first_data_column_position:]
+        )
+
+        # Reassign the categories for any categorical index cols
+        reassign_categories(
+            categories, out._index._data,
+            indices[:first_data_column_position]
+        )
 
+        # Reassign index and column names
         if isinstance(objs[0].columns, pd.MultiIndex):
             out.columns = objs[0].columns
         else:
-            out.columns = unique_columns_ordered_ls
-
-        # Reassign the categories
-        for name, cats in categories.items():
-            out[name] = build_categorical_column(
-                categories=cats,
-                codes=out[name]._column,
-                mask=out[name]._column.mask,
-            )
+            if first_data_column_position <= 1:
+                # If only one index column, reassign its name
+                out._index.name = objs[0]._index.name
+            else:
+                # If more than one index column, reassign MultiIndex names
+                out._index.names = objs[0]._index.names
+            out.columns = names
 
         # TODO (ptaylor):
         # libcudf.nvtx.nvtx_range_pop()
 
-        # Return the columns in the desired out order since this might've
-        # changed if we reassigned categorical categories
-        return out[list(unique_columns_ordered_ls)]
+        return out[list(names)]
```

Comment:
What is in `out` here that we want to select out?

Reply:
`out` is the concatenated DF -- this is selecting columns in the order of `names` in case the columns aren't in that order already.